Experimental/src/Ranges/Range.cs

```diff
@@ -0,0 +1,713 @@
+        where TValue : IComparable<TValue>
+    {
+		public static readonly Range<TValue> Empty = new Range<TValue>(default(TValue), default(TValue),
+            RangeOptions.IsEmpty);
+
+        public static readonly Range<TValue> Full = new Range<TValue>(default(TValue), default(TValue),
+            RangeOptions.None);
+
+        private readonly RangeOptions _options;
+        public readonly TValue End;
+        public readonly TValue Start;
+
+        public Range(TValue start, TValue end, RangeOptions options)
+        {
+            if (options.HasFlag(RangeOptions.IsEmpty))
+            {
+                // clear other flags
+                options = RangeOptions.IsEmpty;
+            }
+
+            if (ReferenceEquals(start, null))
+                options &= ~RangeOptions.HasStart;
+
+            if (ReferenceEquals(end, null))
+                options &= ~RangeOptions.HasEnd;
+
+            if (!options.HasFlag(RangeOptions.HasStart))
+            {
+                start = default(TValue);
+                options &= ~RangeOptions.IncludingStart;
+            }
+
+            if (!options.HasFlag(RangeOptions.HasEnd))
+            {
+                end = default(TValue);
+                options &= ~RangeOptions.IncludingEnd;
+            }
+
+            _options = options;
+
+            if (options.HasFlag(RangeOptions.HasStart) && options.HasFlag(RangeOptions.HasEnd))
+            {
+                var compare = start.CompareTo(end);
+                if (compare > 0)
+                {
+                    // clear including flags before split
+                    _options &= ~(RangeOptions.IncludingStart | RangeOptions.IncludingEnd);
+
+                    Start = end;
+                    End = start;
+                    if (options.HasFlag(RangeOptions.IncludingStart))
+                        _options |= RangeOptions.IncludingEnd;
+                    if (options.HasFlag(RangeOptions.IncludingEnd))
+                        _options |= RangeOptions.IncludingEnd;
+                }
+                else
+                {
+                    Start = start;
+                    End = end;
+
+                    if (compare == 0)
+                    {
+                        if (!(options.HasFlag(RangeOptions.IncludingStart)
+                              || options.HasFlag(RangeOptions.IncludingEnd)))
+                        {
+                            // empty
+                            Start = default(TValue);
+                            End = default(TValue);
+                            _options = RangeOptions.IsEmpty;
+                        }
+                        else
+                        {
+                            // ensure that everyting included
+                            _options |= RangeOptions.IncludingStart | RangeOptions.IncludingEnd;
+                        }
+                    }
+                }
+            }
+            else
```

Comment:
Better to invert the corresponding if to minimize nesting. Resharper should suggest this change.